docs/site/extending/rest-api.md

```diff
@@ -0,0 +1,184 @@
+---
+lang: en
+title: 'Contributing REST API endpoints'
+keywords: LoopBack 4, Extensions, Components
+sidebar: lb4_sidebar
+permalink: /doc/en/lb4/creating-components-rest-api.html
+---
+
+## Overview
+
+As mentioned in [Creating components](../Creating-components.md), components can
+contribute new REST API endpoints by adding Controller classes to the list of
+controllers that should be mounted on the target application.
+
+```ts
+class MyController {
+  @get('/ping')
+  ping() {
+    return {running: true};
+  }
+}
+
+export class MyComponent {
+  constructor() {
+    this.controllers = [MyController];
+  }
+}
+```
+
+This is approach works great when the metadata is hard-coded, e.g. the "ping"
+endpoint path is always `/ping`.
+
+In practice, components often need to allow applications to configure REST API
+endpoints contributed by a component:
+
+- Produce OpenAPI dynamically based on the component configuration, e.g.
+  customize the path of controller endpoints.
+
+- Exclude certain controller endpoints based on configuration. For example, our
+  REST API explorer has a flag `useSelfHostedSpec` that controls whether a new
+  endpoint serving OpenAPI document is added.
+
+## Dynamic OpenAPI metadata
```

Comment:
We could use https://loopback.io/doc/en/lb4/Booting-an-Application.html#boot-an-application-using-component for some of the use cases. I was able to modify the binding key and other metadata for services/controllers/... loaded by booters by override the `app.service()`, `app.controller()` for the subapp.

Reply:
Thank you for the pointer, I agree "booting an app using component" is a concept that can be useful when migrating LB3 extensions to LB4. However, I don't see how to apply it to "Dynamic OpenAPI metadata". I opened a new issue to document "booting an app using component" for extension authors in a holistic way - see https://github.com/strongloop/loopback-next/issues/5422